### PR TITLE
feat: Add timestamp option to set property names

### LINKED ISSUE #142: Feature request: Be able to decide the name of timestamps
Hey, I would like to be able to choose different names for the timestamps. Maybe configuring like this:

```
import { schema } from 'papr';

const schema = schema({
  ...
}, {
  timestamps: {createdAt: '_createdDate', updatedAt: '_updatedDate'}
});
```

### PR DESCRIPTION
**Pull request by joshuat**
Closes #142

We facilitate this change by returning the full `TOptions` from schema construction rather than just `TDefaults`. We can then access the default values from `TOptions` when needed. With full access to the options config we can appropriately se the timestamp property names.

This also adds a `$timestamps` property on the schema return value which is ultimately removed in `updateSchema`. This gives us runtime access to the timestamp property names.

Diff best viewed with --ignore-space-change.

BREAKING CHANGE: Changes the return type for `schema` and the generic arguments for `model` and various utils

### CODE REVIEW

File: src/__tests__/schema.test.ts [hunk 1]
```diff
@@ -120,61 +120,249 @@ describe('schema', () => {
...
+              updatedAt: string;
+            };
+          }
+        ]
+      >(value);
+      expectType<ObjectId>(value[0]?._id);
+      expectType<boolean | undefined>(value[0]?.foo);
+      expectType<Date>(value[0]?._createdDate);
+      expectType<Date>(value[0]?._updatedDate);
+      expectType<typeof value[0]>({
```
Comment (avaly) [inline src/__tests__/schema.test.ts:293]:
```suggestion
      // @ts-expect-error `createdAt` is undefined here
      value[0]?.createdAt;
      // @ts-expect-error `updatedAt` is undefined here
      value[0]?.updatedAt;
      expectType<typeof value[0]>({
```

File: src/__tests__/schema.test.ts [hunk 2]
```diff
@@ -120,61 +120,249 @@ describe('schema', () => {
...
+              createdAt: string;
+            };
+          }
+        ]
+      >(value);
+      expectType<ObjectId>(value[0]?._id);
+      expectType<boolean | undefined>(value[0]?.foo);
+      expectType<Date>(value[0]?._createdDate);
+      expectType<Date>(value[0]?.updatedAt);
+      expectType<typeof value[0]>({
```
Comment (avaly) [inline src/__tests__/schema.test.ts:360]:
```suggestion
      // @ts-expect-error `createdAt` is undefined here
      value[0]?.createdAt;
      expectType<typeof value[0]>({
```

File: src/__tests__/model.test.ts [hunk 3]
```diff
@@ -48,12 +48,34 @@ describe('model', () => {
...
-  type SimpleDefaults = typeof simpleSchema[1];
+  type SimpleOptions = typeof simpleSchema[1];
   type TimestampsDocument = typeof timestampsSchema[0];
-  type TimestampsDefaults = typeof timestampsSchema[1];
-  let simpleModel: Model<SimpleDocument, SimpleDefaults>;
-  let timestampsModel: Model<TimestampsDocument, TimestampsDefaults>;
+  type TimestampsOptions = typeof timestampsSchema[1];
+  type TimestampConfigDocument = typeof timestampConfigSchema[0];
+  type TimestampConfigOptions = typeof timestampConfigSchema[1];
+  let simpleModel: Model<SimpleDocument, SimpleOptions>;
```
Comment (avaly) [inline src/__tests__/model.test.ts:76]:
```suggestion

  let simpleModel: Model<SimpleDocument, SimpleOptions>;
```

File: src/__tests__/model.test.ts [hunk 4]
```diff
@@ -578,6 +784,19 @@ describe('model', () => {
...
 
+    test('with timestamp config', async () => {
+      const results = await timestampConfigModel.find({});
+
+      expectType<TimestampConfigDocument[]>(results);
+
+      if (results.length) {
+        expectType<string>(results[0].foo);
+        expectType<number>(results[0].bar);
+        expectType<Date>(results[0]._createdDate);
```
Comment (avaly) [inline src/__tests__/model.test.ts:795]:
```suggestion
        expectType<Date>(results[0]._createdDate);
        expectType<Date>(results[0]._updatedDate);
```

File: src/__tests__/model.test.ts [hunk 5]
```diff
@@ -1240,6 +1528,32 @@ describe('model', () => {
       expectType<Date>(result.updatedAt);
     });
 
+    test('timestamp config schema', async () => {
```
Comment (avaly) [inline src/__tests__/model.test.ts:1531]: We need another test case for `insertOne()` with the `_createdDate` and `_updatedDate` values populated.

**Review by avaly — DISMISSED** [review]
Thanks for this PR! It's truly amazing!

We'll take a couple more days to test-drive this in our private repo first and I'll follow up with any additional feedback.

**Comment by avaly** [comment]
I don't expect a lot of changes in performance from this PR, but given how many changes it contains, would you mind posting some results from running `yarn benchmark` on `main` vs this PR?

File: src/__tests__/model.test.ts (hunk 4 above)
Comment (joshuat) [reply src/__tests__/model.test.ts:795]: Not sure how I missed this one 🤨

**Comment by joshuat** [comment]
@avaly 

<details>
  <summary>PR benchmark results</summary>
  
```
mongodb.insert ~ 1036.25 ops/sec
papr.insert ~ 1025.58 ops/sec
mongoose.insert ~ 587.50 ops/sec
---
mongodb.find ~ 894.48 ops/sec
papr.find ~ 846.84 ops/sec
mongoose.find ~ 571.47 ops/sec
---
mongodb.update ~ 1096.22 ops/sec
papr.update ~ 1051.26 ops/sec
mongoose.update ~ 942.88 ops/sec
```
</details>


<details>
  <summary>main benchmark results</summary>

```
mongodb.insert ~ 1047.50 ops/sec
papr.insert ~ 1041.87 ops/sec
mongoose.insert ~ 578.04 ops/sec
---
mongodb.find ~ 900.91 ops/sec
papr.find ~ 878.20 ops/sec
mongoose.find ~ 563.92 ops/sec
---
mongodb.update ~ 1004.04 ops/sec
papr.update ~ 975.30 ops/sec
mongoose.update ~ 883.93 ops/sec
```
</details>

As expected, nothing out of the ordinary 👍

**Comment by joshuat** [comment]
ping @avaly - any feedback from internal testing? I've just rebased this to catch up with recent changes

**Comment by avaly** [comment]
@joshuat Sorry for the delay, we've been blocked a bit on this testing due to `mongodb` v4.9.0 which introduced a regression. Now that v4.9.1 is out, I'll give this a try next week.

File: src/schema.ts [hunk 6]
```diff
@@ -1,10 +1,23 @@
...
 
-interface SchemaOptions<TProperties> {
+export type SchemaTimestampOptions =
+  | boolean
+  | RequireAtLeastOne<{
+      createdAt?: string;
+      updatedAt?: string;
+    }>;
+
+export interface SchemaOptions<TProperties> {
```
Comment (avaly) [inline src/schema.ts:18]: Can we please export this type in the `index.ts`?

This is the first issue we ran into while trying this PR in our repo.

Comment (joshuat) [reply src/schema.ts:18]: Done!

**Pull request merged**